### PR TITLE
parser: Fix #288

### LINKED ISSUE #288: `s.SetText(yylex.(*lexer).src[yyS[yypt - 1].col-1:yyS[yypt].col-1])` panics for sub select.
```
SELECT stuff.id 
FROM stuff 
WHERE stuff.value >= ALL (SELECT stuff.value 
FROM stuff)
```

If the query is not in one line for subselect, parser will panic.

### PR DESCRIPTION
**Pull request by ngaut**
Fix get text of subselect statement.

The solution is not pretty. We may need to change goyacc for getting origin text of expressions.

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -2982,7 +2982,28 @@ SubSelect:
 	'(' SelectStmt ')'
 	{
 		s := $2.(*stmts.SelectStmt)
-		s.SetText(yylex.(*lexer).src[yyS[yypt - 1].col-1:yyS[yypt].col-1])
+		src := yylex.(*lexer).src
+		lines := yyS[yypt-1].line-1
+		pos := 0
+		if lines > 0 {
```
Comment (shenli) [inline parser/parser.y]: I can not understand this. Any comment?

Comment (ngaut) [inline parser/parser.y]: yyS and yypt is internal variables for yacc.

File: parser/parser.y [hunk 2]
```diff
@@ -2982,7 +2982,28 @@ SubSelect:
...
+					cnt++
+				}
+			}
+		}
+
+		base := pos+yyS[yypt-1].col-1
+		end := yyS[yypt].col-1
+		if end < base {
+			s.SetText(src[base:])
+		} else{ // when yypt in new line
```
Comment (c4pt0r) [inline parser/parser.y]: code style

File: parser/parser_test.go [hunk 3]
```diff
@@ -88,6 +88,10 @@ func (s *testParserSuite) TestParser0(c *C) {
 		{"CREATE TABLE foo (a SMALLINT UNSIGNED, b INT UNSIGNED) // foo", true},
 		{"CREATE TABLE foo (a SMALLINT UNSIGNED, b INT UNSIGNED) /* foo */", true},
 		{"CREATE TABLE foo /* foo */ (a SMALLINT UNSIGNED, b INT UNSIGNED) /* foo */", true},
+		{`SELECT stuff.id 
+		FROM stuff 
+		WHERE stuff.value >= ALL (SELECT stuff.value
```
Comment (c4pt0r) [inline parser/parser_test.go]: strange line break here

File: parser/parser.y (hunk 1 above)
Comment (c4pt0r) [inline parser/parser.y]: Please add a comment here. For future maintenance.

File: parser/parser_test.go (hunk 3 above)
Comment (ngaut) [inline parser/parser_test.go]: The test is used for testing multiple lines statement

**Comment by ngaut** [comment]
PTAL @c4pt0r @shenli

File: parser/parser.y [hunk 4]
```diff
@@ -2982,7 +2982,29 @@ SubSelect:
...
 	{
 		s := $2.(*stmts.SelectStmt)
-		s.SetText(yylex.(*lexer).src[yyS[yypt - 1].col-1:yyS[yypt].col-1])
+		src := yylex.(*lexer).src
+		// See the implemention of yyParse function
+		lines := yyS[yypt-1].line-1
+		pos := 0
+		if lines > 0 {
+			cnt := 0
+			for ; pos < len(src); pos++{
```
Comment (siddontang) [inline parser/parser.y]: use IndexByte is much faster than looping over the bytes in src.

File: parser/parser.y [hunk 5]
```diff
@@ -2982,7 +2982,29 @@ SubSelect:
...
+					cnt++
+				}
+			}
+		}
+
+		base := pos+yyS[yypt-1].col-1
+		end := yyS[yypt].col-1
+		if end < base {
+			s.SetText(src[base:])
+		} else { // when yypt in new line
```
Comment (siddontang) [inline parser/parser.y]: should comment be put in a new line?

File: parser/parser.y (hunk 4 above)
Comment (ngaut) [inline parser/parser.y]: Not sure in this case, simply loop is clean and clear.

**Comment by siddontang** [comment]
```
tidb> select (
   -> select 1
   -> );
+--------------+
| (select 1
)) |
+--------------+
| 1            |
+--------------+
```

A redundant  ')'

**Comment by ngaut** [comment]
Good catch, @siddontang

File: parser/parser.y (hunk 4 above)
Comment (siddontang) [inline parser/parser.y]: http://play.golang.org/p/jH8auHsni_

In my computer, IndexBytes is nearly 2x faster than looping over.

**Comment by ngaut** [comment]
@shenli @siddontang @c4pt0r 
Introduce a new pretty way to handle subselect.  And we are able to handle expression too :)

**Comment by shenli** [comment]
Please add an unittest to check subselect statement.Text.

**Comment by shenli** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**